### PR TITLE
Track proposal expiry

### LINKED ISSUE #3642: Track Proposal Procedure Expiry
GovAction must expire after the number of epochs specified in PParams.

It is also important not to affect expiry of current proposals, when `govActionExpiration` param is updated. For this we need to record the epoch when action expires in GovActionsState whenever ProposalProcedure is being proposed

### PR DESCRIPTION
**Pull request by aniketd**
# Description

Closes #3642 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs [hunk 1]
```diff
@@ -59,8 +59,9 @@ import NoThunks.Class (NoThunks (..))
 import Validation (failureUnless)
 
 data GovEnv era = GovEnv
-  { teTxId :: !(TxId (EraCrypto era))
-  , teEpoch :: !EpochNo
+  { geTxId :: !(TxId (EraCrypto era))
+  , geEpoch :: !EpochNo
+  , geGovActionExpiry :: !EpochNo
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs]: I think it is better to just pass all of PParams, since we'll need some of the other ones in the GOV rule in the future. Namely `minCommitteSize` and `govActionDeposit`
```suggestion
  , gePParams :: !PParams era
```

File: libs/cardano-ledger-pretty/cardano-ledger-pretty.cabal [hunk 2]
```diff
@@ -39,7 +39,7 @@ library
         cardano-ledger-alonzo >=1.2,
         cardano-ledger-babbage >=1.1,
         cardano-ledger-byron,
-        cardano-ledger-conway ^>=1.7,
+        cardano-ledger-conway >=1.7,
```
Comment (lehins) [inline libs/cardano-ledger-pretty/cardano-ledger-pretty.cabal]: That bound is invalid. You made a breaking change that `cardano-ledger-pretty` now depends on.
```suggestion
        cardano-ledger-conway ^>=1.8,
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs [hunk 3]
```diff
@@ -217,11 +213,13 @@ ratifyTransition = do
...
+          -- This action hasn't been ratified yet. Process the remaining actions.
           st' <- trans @(ConwayRATIFY era) $ TRC (env, st, RatifySignal sigs)
-          if expired
-            then -- Action expired, do not include it in the next epoch
+          -- Finally, filter out actions that are not processed.
+          if gasExpiresAfter >= reCurrentEpoch
+            then -- Action expires after current Epoch. Remove it.
               pure $ st' {rsRemoved = act :<| rsRemoved}
-            else -- Include this action in the next epoch
+            else -- Action neither expires, nor is enacted, yet. Carry it over to the next Epoch.
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: This comment can be removed, since `rsFuture` will be removed in another PR
```suggestion
            else
```

File: eras/conway/impl/CHANGELOG.md [hunk 4]
```diff
@@ -1,9 +1,12 @@
...
 
-## 1.7.0.1
-
-*
+## 1.8.0.0
 
+* Implement expiry for governance proposals #3664
+  * Update `ppGovActionExpiration` to be an `EpochNo`
+  * Add `gasExpiresAfter :: !EpochNo` to `GovActionState`
+  * Add `geGovActionExpiry` to `GovEnv` from the `PParams`
```
Comment (lehins) [inline eras/conway/impl/CHANGELOG.md]:
```suggestion
  * Add `gePParams` to `GovEnv`.
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs [hunk 5]
```diff
@@ -217,11 +213,13 @@ ratifyTransition = do
                   }
           trans @(ConwayRATIFY era) $ TRC (env, st', RatifySignal sigs)
         else do
+          -- This action hasn't been ratified yet. Process the remaining actions.
           st' <- trans @(ConwayRATIFY era) $ TRC (env, st, RatifySignal sigs)
-          if expired
-            then -- Action expired, do not include it in the next epoch
+          -- Finally, filter out actions that are not processed.
+          if gasExpiresAfter >= reCurrentEpoch
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: For consistency with DRep expiration, the spec and the name of the record (After) :slightly_smiling_face: 

Also, as @teodanciu pointed out, expiration should be smaller than the current epoch in order for it to expire. It starts out as the epoch in the future (or current epoch, if the `govActionExpiration` pparam is set to 0)
```suggestion
          if gasExpiresAfter < reCurrentEpoch
```

**Review by lehins — CHANGES_REQUESTED** [review]
Looks good. Few important changes are needed though

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs (hunk 5 above)
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: Wait, so if the currentEpoch is smaller than the expiration Epoch, this means that it's expired?  Is this right?

Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: You are absolutely right :smile: Updated my suggestion.

File: libs/cardano-ledger-pretty/cardano-ledger-pretty.cabal (hunk 2 above)
Comment (aniketd) [reply libs/cardano-ledger-pretty/cardano-ledger-pretty.cabal]: Oh yes! What I did is wrong :+1: Thanks!

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs (hunk 5 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: :face_with_head_bandage: This mistake made me feel a bit sick :face_with_spiral_eyes:. Thanks for catching this!

Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs]: Thanks for the awesome catch :raised_hands:

**Review by lehins — APPROVED** [review]
:+1:
